code/modules/projectiles/guns/energy.dm

```diff
@@ -411,3 +411,129 @@
+						/obj/item/attachable/attached_gun/shotgun,
+						/obj/item/attachable/scope/mini)
+	flags_gun_features = GUN_AUTO_EJECTOR|GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_ENERGY|GUN_AMMO_COUNTER
+	attachable_offset = list("muzzle_x" = 32, "muzzle_y" = 18,"rail_x" = 12, "rail_y" = 23, "under_x" = 23, "under_y" = 15, "stock_x" = 22, "stock_y" = 12)
+
+	accuracy_mult_unwielded = 0.5 //Heavy and unwieldy; you don't one hand this.
+	scatter_unwielded = 100 //Heavy and unwieldy; you don't one hand this.
+	damage_falloff_mult = 0.25
+	fire_delay = 3
+	var/current_mode = 1
+	var/list/datum/lasrifle/base/mode_list = list(
+						/datum/lasrifle/base/standard,
+						/datum/lasrifle/base/auto,
+						/datum/lasrifle/base/overcharge,
+						/datum/lasrifle/base/heat,
+						/datum/lasrifle/base/spread)
+
+/datum/lasrifle/base
+	var/charge_cost = 0
+	var/ammo = null
+	var/fire_delay = 0
+	var/fire_sound = null
+	var/message_to_user = ""
+	var/fire_mode = GUN_FIREMODE_SEMIAUTO
+	var/icon_state = "tx73"
+
+/datum/lasrifle/base/standard
+	charge_cost = 20
+	ammo = /datum/ammo/energy/lasgun/M43
+	fire_delay = 3
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasrifle's charge mode to standard fire."
+	icon_state = "tx73"
+
+/datum/lasrifle/base/auto
+	charge_cost = 20
+	ammo = /datum/ammo/energy/lasgun/M43
+	fire_delay = 2
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasrifle's charge mode to automatic fire."
+	fire_mode = GUN_FIREMODE_AUTOMATIC
+	icon_state = "tx73_auto"
+
+/datum/lasrifle/base/overcharge
+	charge_cost = 80
+	ammo = /datum/ammo/energy/lasgun/M43/overcharge
+	fire_delay = 10
+	fire_sound = 'sound/weapons/guns/fire/plasma_precision_3.ogg'
+	message_to_user = "You set the Lasrifle's charge mode to overcharge."
+	icon_state = "tx73_overcharge"
+
+/datum/lasrifle/base/heat
+	charge_cost = 80
+	ammo = /datum/ammo/energy/lasgun/M43/heat
+	fire_delay = 8
+	fire_sound = 'sound/weapons/guns/fire/laser.ogg'
+	message_to_user = "You set the Lasrifle's charge mode to wave heat."
+	icon_state = "tx73_heat"
+
+/datum/lasrifle/base/spread
+	charge_cost = 80
+	ammo = /datum/ammo/energy/lasgun/M43/blast
+	fire_delay = 8
+	fire_sound = 'sound/weapons/guns/fire/laser_rifle_2.ogg'
+	message_to_user = "You set the Lasrifle's charge mode to spread."
+	icon_state = "tx73_spread"
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/unique_action(mob/user)
+	return switch_modes(user)
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle/proc/switch_modes(mob/user)
+	var/max_overcharge_mode = length(mode_list)
+	if(current_mode >= max_overcharge_mode)
+		current_mode = 1
+	else
+		current_mode += 1
+	
+	playsound(user, 'sound/weapons/emitter.ogg', 5, 0, 2)
+	charge_cost = initial(mode_list[current_mode].charge_cost)
+	ammo = GLOB.ammo_list[initial(mode_list[current_mode].ammo)]
+	fire_delay = initial(mode_list[current_mode].fire_delay)
+	fire_sound = initial(mode_list[current_mode].fire_sound)
```

Comment:
Why the use of `initial` all over?

Reply:
No particular reason for the method.

I need it to get the relevant information from each datum in the list, and `initial` does that.

Reply:
initial is correct here to avoid having to instance the datums